cognitive/src/main/scala/com/microsoft/ml/spark/cognitive/FormRecognizer.scala

```diff
@@ -0,0 +1,634 @@
+        it.next()
+      } else {
+        throw new TimeoutException(
+          s"Querying for results did not complete within $maxTries tries")
+      }
+    } else {
+      response
+    }
+  }
+
+  override protected def responseDataType: DataType = TrainCustomModelResponse.schema
+
+}
+
+object AnalyzeCustomForm extends ComplexParamsReadable[AnalyzeCustomForm] {
+  def flattenReadResults(inputCol: String, outputCol: String): UDFTransformer = {
+    val fromRow = AnalyzeCustomFormResponse.makeFromRowConverter
+    def extractText(lines: Array[ReadLine]): String = {
+      lines.map(_.text).mkString(" ")
+    }
+    new UDFTransformer()
+      .setUDF(UDFUtils.oldUdf(
+        { r: Row =>
+          Option(r).map(fromRow).map(
+            _.analyzeResult.readResults.map(_.lines.map(extractText).mkString("")).mkString(" ")).mkString("")
+        },
+        StringType))
+      .setInputCol(inputCol)
+      .setOutputCol(outputCol)
+  }
+
+  def flattenPageResults(inputCol: String, outputCol: String): UDFTransformer = {
+    val fromRow = AnalyzeCustomFormResponse.makeFromRowConverter
+    def extractTableText(pageResults: Seq[CustomFormPageResult]): String = {
+      pageResults.map(_.tables.map(_.cells.map(_.text).mkString(" | ")).mkString("\n")).mkString("\n\n")
+    }
+    def generateKeyValuePairs(keyValuePair: KeyValuePair): String = {
+      "key: " + keyValuePair.key.text + " value: " + keyValuePair.value.text
+    }
+    def extractKeyValuePairs(pageResults: Seq[CustomFormPageResult]): String = {
+      pageResults.map(_.keyValuePairs.map(generateKeyValuePairs).mkString("\n")).mkString("\n\n")
+    }
+    def extractAllText(pageResults: Seq[CustomFormPageResult]): String = {
+      "KeyValuePairs: " + extractKeyValuePairs(pageResults) + "\n\n\n" + "Tables: " + extractTableText(pageResults)
+    }
+    new UDFTransformer()
+      .setUDF(UDFUtils.oldUdf(
+        { r: Row =>
+          Option(r).map(fromRow).map(
+            _.analyzeResult.pageResults.map(extractAllText).mkString(" "))
+        },
+        StringType))
+      .setInputCol(inputCol)
+      .setOutputCol(outputCol)
+  }
+
+  def flattenDocumentResults(inputCol: String, outputCol: String): UDFTransformer = {
+    val fromRow = AnalyzeCustomFormResponse.makeFromRowConverter
+    def extractFields(documentResults: Seq[DocumentResult]): String = {
+      documentResults.map(_.fields).mkString("\n")
+    }
+    new UDFTransformer()
+      .setUDF(UDFUtils.oldUdf(
+        { r: Row =>
+          Option(r).map(fromRow).map(
+            _.analyzeResult.documentResults.map(extractFields).mkString("")).mkString("")
+        },
+        StringType))
+      .setInputCol(inputCol)
+      .setOutputCol(outputCol)
+  }
+}
+
+class AnalyzeCustomForm(override val uid: String) extends FormRecognizerBase(uid) with BasicLogging {
+  logClass()
+
+  def this() = this(Identifiable.randomUID("AnalyzeCustomForm"))
+
+  def setLocationAndModelId(loc: String, modelId: String): this.type =
```

Comment:
perhaps we should make a separate (service) parameter called modelID so that the setLocation in the base cog services works as expected and users can use many diff custom models with a single call if desired